core/jobs/types/model_property.py

```diff
@@ -109,62 +128,92 @@ def yield_value_from_model(self, model):
         else:
             yield value
 
-    def _to_model_class(self):
+    def _to_model_class(self) -> Type[base_models.BaseModel]:
         """Returns the model class associated with this instance.
 
         Returns:
             type(BaseModel). The model type.
         """
-        return job_utils.get_model_class(self._model_kind)
-
-    def _to_property(self):
+        # Here, get_model_class() returns the value of type
+        # Type[datastore_services.Model] but the expected return value from
+        # this method is Type[base_models.BaseModel]. So, that's why we used
+        # cast here.
+        return cast(
+            Type[base_models.BaseModel],
+            job_utils.get_model_class(self._model_kind)
+        )
+
+    def _to_property(
+        self
+    ) -> Union[
+        datastore_services.Property,
+        Callable[[base_models.BaseModel], str]
+    ]:
         """Returns the Property object associated with this instance.
 
         Returns:
             *. A property instance.
         """
-        return getattr(self._to_model_class(), self._property_name)
-
-    def _is_repeated_property(self):
+        # Using a cast here because the return value of getattr() method is
+        # dynamic and mypy will assume it to be Any type otherwise.
+        return cast(
+            Union[
+                datastore_services.Property,
+                Callable[[base_models.BaseModel], str]
+            ],
+            getattr(self._to_model_class(), self._property_name)
+        )
+
+    def _is_repeated_property(self) -> bool:
         """Returns whether the property is repeated.
 
         Returns:
             bool. Whether the property is repeated.
         """
-        return self._property_name != 'id' and self._to_property()._repeated # pylint: disable=protected-access
-
-    def __getstate__(self):
+        model_property = self._to_property()
+        # We used assert here, because this method checks repetition only
+        # for those properties which belongs to datastore_services.Property
+        # and id is not one of them. Because id is a python property and it
+        # does not contain '_repeated' attribute.
+        assert isinstance(model_property, datastore_services.Property)
```

Comment:
is this actually an instance of `datastore_services.Property`?  or [`ModelProperty`](https://github.com/oppia/oppia/blob/2a115bbf857c790b7433daba21a6b27d32af0f51/core/jobs/types/model_property.py#L29). I am not sure.

This statement is breaking the backend_tests

Reply:
Here, `model_property` can be an instance of both `datastore_services.Property` or python property. So, if we look closely at the code of this function we can see that this function only checks the repetition of a property if it is not the `id` property ( which is the only python property this class `ModelProperty` can accept. ). Due to this, MyPy assumes `Union[datastore_services.Property, Callable[[base_models.BaseModel], str]]` is the type of `model_property`. But we know only properties of type `datastore_services.Property` can contain `_repeated` attribute. So to remove the conflict and narrow down the union type I used assert statement.

But this is fixed now, by using the if clause to narrow down the union type.

Reply:
Why it can be both `datastore_services.Property` and `Callable[[base_models.BaseModel], str]` here?

Reply:
Ahh, sorry. I mean to say `model_property` can be of either `datastore_services.Property` type or `Callable[[base_models.BaseModel], str]` type.